repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/apivalidator/ApiValidatorMultiTenantTest.groovy

```diff
@@ -0,0 +1,475 @@
+
+    static def mapHeaderValue = base64EncodeUtf8("""{
+      |   "1" : ["a:admin","foo","bar"],
+      |   "2" : ["a:creator", "foo", "a:observer"],
+      |   "3" : ["a:updater", "bar", "biz", "a:creator"],
+      |   "4" : ["a:observer"],
+      |   "5" : ["a:admin", "bar", "biz", "a:creator"],
+      |   "6" : ["biz", "baz"],
+      |   "7" : ["role with spaces", "biz"],
+      |   "8" : ["a:patcher"]
+      |}
+      |""".stripMargin())
+
+    static def observerRequests = [
+        ["GET", "/v1/resource"]
+    ]
+
+    static def creatorRequests = [
+        ["POST", "/v1/resource"],
+        ["POST", "/v1/resource/other"]
+    ]
+    static def updaterRequests = [
+        ["PUT", "/v1/resource"],
+        ["PUT", "/v1/resource/other"]
+    ]
+    static def adminOnlyRequests = [
+        ["DELETE", "/v1/resource"],
+        ["DELETE", "/v1/resource/other"]
+    ]
+
+
+    def setupSpec() {
+        deproxy = new Deproxy()
+        deproxy.addEndpoint(properties.targetPort)
+
+        def params = properties.getDefaultTemplateParams()
+        repose.configurationProvider.applyConfigs("common", params)
+        repose.configurationProvider.applyConfigs("features/filters/apivalidator/multitenant", params)
+        repose.start()
+
+        repose.waitForNon500FromUrl(reposeEndpoint)
+    }
+
+    @Unroll("Should succeed on GET /v1/resource/other when appropriate headers are set (#tenants)")
+    def "Should succeed when appropriate headers are set"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint/v1/resource/other",
+            method: "GET",
+            headers: [
+                "X-Tenant"   : tenants,
+                "X-Other"    : "other",
+                (ROLES)      : "foo,bar",
+                "X-Map-Roles": mapHeaderValue])
+
+        then:
+        mc.getReceivedResponse().getCode() as Integer == SC_OK
+        mc.handlings.size() == 1
+        List<String> relevantRolesValues = mc.handlings[0].request.headers.findAll(RELEVANT_ROLES)
+        relevantRolesValues.containsAll(["foo", "bar"])
+
+        where:
+        tenants << ["1,5", "4", "4,2", "1,2,3", "5,2", "1,3", "5,3", "1", "1,5,7", "1,5,6"]
+    }
+
+    @Unroll("Should succeed on #method #url when appropriate headers are set (1, 5)")
+    def "Should succeed when appropriate headers are set (1, 5)"() {
+        when:
+        MessageChain mc = deproxy.makeRequest(
+            url: "$reposeEndpoint$url",
+            method: method,
+            headers: [
+                "X-Tenant"   : "1,5",
+                "X-Other"    : "other",
+                (ROLES)      : "foo,bar",
+                "X-Map-Roles": mapHeaderValue])
+
+        then:
+        mc.getReceivedResponse().getCode() as Integer == SC_OK
+        mc.handlings.size() == 1
+        List<String> relevantRolesValues = mc.handlings[0].request.headers.findAll(RELEVANT_ROLES)
+        relevantRolesValues.contains("a:admin/{X-TENANT}")
```

Comment:
Huh? Relevant roles contain the literal `{X-TENANT}`? Shouldn't that be replaced with the parameterized request tenant?

Reply:
I'm told this is what `api-checker` does. Weird. It seems like if we had the actual tenant we would be able to scope down the request tenants even more in the `TenantCullingFilter`. That is, having the exact relevant role(s)/tenant(s) would be nice data to have.

Reply:
@dmnjohns  :  No, it is not replaced with the tenant value -- it is the name of the parameter.  Consider that it may actually be multiple values.  `TenantCullingFilter` shouldn't scope based on the tenant, it should scope based on roles in the relevant roles header. In fact there are use cases that would break if you scoped down to the tenant first -- TenantCulling should just cull based on roles.

Reply:
... also `TenantCullingFilter` should ignore anything after `/` in the `X-RELEVANT-ROLES` header.

Reply:
Right, that is exactly how the `TenantCullingFilter` works right now.

I suppose what I was saying is that while the `X-Tenant` param may be multi-valued, when we get to the point that we have relevant roles, we should know exactly what role(s) on what tenant(s) allowed the request through, right? So we should be able to forward that information along for whatever purpose, such as additional culling. So say I have a wadl with:
```
<method name="POST" rax:roles="a:creator/{X-TENANT}">
               <request>
                   <param name="X-TENANT"
                          style="header"
                          required="true"
                          type="xsd:int"
                          repeating="true"
                          rax:isTenant="true"/>
```

That snippet essentially says that to `POST`, a user must have the role `a:creator` on at least one of the tenants we parse from the `X-Tenant` header, right? Well, if that header is `X-Tenant: tenant1, tenant2, tenant3`, and if the user has `a:creator` on `tenant1` and `tenant2` but not on `tenant3`, then rather than saying relevant roles are `a:creator/{X-TENANT}` could we not say they are `a:creator/tenant1, a:creator/tenant2`? It's just further specification. I don't actually know that such specification would serve a purpose, but...

So in the case with the `TenantCullingFilter`, it wouldn't be scoping based on tenant per se, it would be scoping based on roles knowing specifically what roles were relevant rather than vaguely what roles were relevant. Using the example above, then, we could strip `tenant3` out of the request.

Reply:
Nope, actually in the example above if the header on the request is `X-TENANT: 1, 2, 3` (note `tenant1` doesn't validate against int).  And the user only has `a:creator` on `1` and `2` but not on `3`. Then the request would fail, since one of the tenants is not allowed the post.  The difference between `a:creator` on  Relevant Role header and `a:creator/{X-TENANT}` is simply stated that we made the tenant check.  If you want to allow access based on *any* tenant having the role then you should say `rax:roles="a:creator"` makes sense?

Reply:
Fair enough, I should have used ints. So it's an *AND* rather than an *OR* for the tenanted role check. That makes way more sense to me from a multi-tenant perspective. I think my confusion arose from the fact that all of the other role requirements seem to be *OR*s, right? That is, `rax:roles="role1 role2"` would require `role1` or `role2`? And same with ancestral role requirements (like `rax:roles="role1"` on `/` and `rax:roles="role2"` on `/foo`)?

Anyway, in that case, we're good. Effectively, `relevantRole/{X-Tenant}` (or whatever the parameter happens to be named) denotes that `relevantRole` was relevant for *ALL* tenants taken from the `X-Tenant` parameter.

Reply:
It's an `OR` for roles and an `AND` for tenants.   Or in other words any one of these roles on all of the tenants.  Soon, we need to do an `AND` for tenants -- I'm not sure what the syntax for that will look like...